### ISSUE TITLE
Efficient batched inference

### ISSUE THREAD

**Issue by Dan-wanna-M** [Dan-wanna-M/formatron#3]
While we support batched inference like other constrained decoding libraries, the current implementation can be parallelized further. In particular, we can mask logits in batch and run several `kbnf` engine in parallel.

- [ ] Add batching API to `kbnf` and expose relevant API
- [ ] Test scalability on huggingface
- [ ] Request `batch_logits_processors` feature on other integrations
  - [ ] vllm
  - [ ] exllamav2

**Comment by turboderp** [comment]
I'm not sure what the control flow is, but at least ExLlamaV2 does multithreaded sampling for batches, so if you can disable the GIL during `compute_allowed_tokens` that should go a long way.

Batching logits for processing would be problematic since ExLlamaV2 allows for concurrent jobs with completely different sampling/filtering settings.

**Comment by turboderp** [comment]
I'm not sure how it translates to Rust, but it's described [here](https://docs.python.org/3/c-api/init.html#releasing-the-gil-from-extension-code) for (C++) extension functions.

Basically wrap a block of code in `Py_BEGIN_ALLOW_THREADS` and `Py_END_ALLOW_THREADS`, and as long as it isn't manipulating any Python objects it will allow other Python threads to continue running while the extension code does some blocking I/O or computation.

**Comment by Dan-wanna-M** [comment]
> I'm not sure how it translates to Rust, but it's described [here](https://docs.python.org/3/c-api/init.html#releasing-the-gil-from-extension-code) for (C++) extension functions.
> 
> Basically wrap a block of code in `Py_BEGIN_ALLOW_THREADS` and `Py_END_ALLOW_THREADS`, and as long as it isn't manipulating any Python objects it will allow other Python threads to continue running while the extension code does some blocking I/O or computation.

Got it,  I need to check how to do this in pyo3.
